### PR TITLE
Plane: Add VTOL descent rate, convert existing rate to climb rate only

### PR DESCRIPTION
**Pull request by Hwurzburg**
Co-authored-by: Reko Meriö K9260@student.jamk.fi
designate existing pilot climb rate for ascent only
add pilot descent rate max param

### CODE REVIEW

File: ArduPlane/quadplane.cpp [hunk 1]
```diff
@@ -39,14 +39,24 @@ const AP_Param::GroupInfo QuadPlane::var_info[] = {
...
+    AP_GROUPINFO("VELZ_MAX", 18, QuadPlane, pilot_velocity_z_max_up, 250),
+   
+   // @Param: VELZ_MAX_DN
+    // @DisplayName: Pilot maximum vertical speed down
+    // @Description: The maximum vertical velocity the pilot may request in cm/s going down
+    // @Units: cm/s
+    // @Range: 50 500
+    // @Increment: 10
+    // @User: Standard
+    AP_GROUPINFO("VELZ_MAX_DN", 60, QuadPlane, pilot_velocity_z_max_dn, 250),
```
Comment (IamPete1) [inline ArduPlane/quadplane.cpp:57]: we should default to 0 being the symmetric with VELZ_MAX, same as copter.

Also it should be down at the bottom in the second group

File: ArduPlane/quadplane.cpp [hunk 2]
```diff
@@ -39,14 +39,24 @@ const AP_Param::GroupInfo QuadPlane::var_info[] = {
     AP_SUBGROUPPTR(pos_control, "P", 17, QuadPlane, AC_PosControl),
 
     // @Param: VELZ_MAX
```
Comment (IamPete1) [inline ArduPlane/quadplane.cpp:41]: It is safe to just change it to `VELZ_MAX_UP`, not sure if its better to or not tho

**Review by IamPete1 — CHANGES_REQUESTED** [review]

File: ArduPlane/quadplane.cpp (hunk 1 above)
Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp:57]: velz max is 250 by default, right? and I thought to keep the params listed together you could do this....its done for the frame and type ..

File: ArduPlane/quadplane.cpp (hunk 2 above)
Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp:41]: wanted to allow people who have changed the value from defualt to still use it when copying params back into a new firmware load... or a new board....thought changing the param name would prevent that

File: ArduPlane/quadplane.cpp (hunk 1 above)
Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp:57]: will do the helper..thanks

Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp]: done

File: ArduPlane/quadplane.cpp (hunk 2 above)
Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp:41]: discuss at devcall

**Comment by Hwurzburg** [comment]
climbrate

File: ArduPlane/quadplane.cpp (hunk 2 above)
Comment (WickedShell) [reply ArduPlane/quadplane.cpp:41]: If one is down, the other should be up, otherwise it's just confusing, IMO.

File: ArduPlane/qautotune.cpp [hunk 3]
```diff
@@ -41,7 +41,7 @@ void QAutoTune::get_pilot_desired_rp_yrate_cd(float &des_roll_cd, float &des_pit
 
 void QAutoTune::init_z_limits()
 {
-    plane.quadplane.pos_control->set_max_speed_z(-plane.quadplane.pilot_velocity_z_max, plane.quadplane.pilot_velocity_z_max);
+    plane.quadplane.pos_control->set_max_speed_z(-plane.quadplane.pilot_velocity_z_max_up, plane.quadplane.get_pilot_velocity_z_max_dn());
```
Comment (IamPete1) [inline ArduPlane/qautotune.cpp:44]:
```suggestion
    plane.quadplane.pos_control->set_max_speed_z(-plane.get_pilot_velocity_z_max_dn(), plane.quadplane.pilot_velocity_z_max_up);
```

File: ArduPlane/quadplane.cpp (hunk 2 above)
Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp:41]: could parameter conversion somehow be used to convert Q_VELZ_MAX, someone tries to set that from an old param file into UP and DN of that value, allowing old param files to work with the new code that changes VELX_MAX to MX_UP?

Comment (WickedShell) [reply ArduPlane/quadplane.cpp:41]: Nope. We don't have any parameter change hooks to do something like that. If we did we'd already have done a lot more sanity checking on some params when people try to set them.

File: ArduPlane/qautotune.cpp (hunk 3 above)
Comment (Hwurzburg) [reply ArduPlane/qautotune.cpp:44]: oops...

File: ArduPlane/quadplane.cpp [hunk 4]
```diff
@@ -333,7 +342,7 @@ const AP_Param::GroupInfo QuadPlane::var_info[] = {
     // @Bitmask: 0:Level Transition,1:Allow FW Takeoff,2:Allow FW Land,3:Vtol Takeoff Frame,4:Use FW Approach,5:Use QRTL,6:Use Governor,7:Force Qassist,8:Mtrs_Only_Qassist,9:Airmode_On_Arm,10:Disarmed Yaw Tilt,11:Delay Spoolup,12:disable Qassist based on synthetic airspeed,13:Disable Ground Effect Compensation,14:Ignore forward flight angle limits in Qmodes
     AP_GROUPINFO("OPTIONS", 58, QuadPlane, options, 0),
 
-    AP_SUBGROUPEXTENSION("",59, QuadPlane, var_info2),
+    AP_SUBGROUPEXTENSION("",60, QuadPlane, var_info2),  // 59 is used above
```
Comment (IamPete1) [inline ArduPlane/quadplane.cpp:345]: I think this has to stay at 59, and sneak the new one in at 60.

**Comment by peterbarker** [comment]
Would be nice to see an autotest for this.

**Comment by robustini** [comment]
This is absolutely useful, after all it has already been present in the Copter for years.
It is correct that the two speeds are separated, since during the ascent we can exaggerate, but not in the descent.

File: ArduPlane/quadplane.cpp (hunk 4 above)
Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp:345]: compiles fine either way...my understanding (could be wrong) and that the second group all resides in that  AP_SUBGROUPEXTENSION at the first groups #59 param slot in the tree...and that 63 params are allowed in that first group..so it seems like that param number should move down and additions to the first group be above it....
..whether you move the slot or not does not seem to impact the compile...maybe something else

and why add in the first group and not the second? to allow it to be put next to its associated param....although GCS sort alphabetically, the parm list in the wiki is generated in order.

Comment (IamPete1) [reply ArduPlane/quadplane.cpp:345]: I think you will loose all the saved param for the second group if you move it. To test you would have to change them on master and then load this branch and compare.

Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp:345]: since it compiles either way, I will not move it....

Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp]: done

File: ArduPlane/quadplane.cpp [hunk 5]
```diff
@@ -333,7 +342,7 @@ const AP_Param::GroupInfo QuadPlane::var_info[] = {
     // @Bitmask: 0:Level Transition,1:Allow FW Takeoff,2:Allow FW Land,3:Vtol Takeoff Frame,4:Use FW Approach,5:Use QRTL,6:Use Governor,7:Force Qassist,8:Mtrs_Only_Qassist,9:Airmode_On_Arm,10:Disarmed Yaw Tilt,11:Delay Spoolup,12:disable Qassist based on synthetic airspeed,13:Disable Ground Effect Compensation,14:Ignore forward flight angle limits in Qmodes
     AP_GROUPINFO("OPTIONS", 58, QuadPlane, options, 0),
 
-    AP_SUBGROUPEXTENSION("",59, QuadPlane, var_info2),
+    AP_SUBGROUPEXTENSION("",59, QuadPlane, var_info2),  // 60 is used above
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:345]:
```suggestion
    AP_SUBGROUPEXTENSION("",59, QuadPlane, var_info2),
    
    // 60 is used above for VELZ_MAX_DN
```

File: ArduPlane/quadplane.cpp [hunk 6]
```diff
@@ -3583,4 +3593,14 @@ bool QuadPlane::show_vtol_view() const
...
 }
 
+// return the PILOT_VELZ_MAX_DN value if non zero, otherwise returns the PILOT_VELZ_MAX value.
+uint16_t QuadPlane::get_pilot_velocity_z_max_dn() const
+{
+    if (pilot_velocity_z_max_dn == 0) {
+        return abs(pilot_velocity_z_max_up);
+    } else {
+        return abs(pilot_velocity_z_max_dn);
+    }
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:3603]:
```suggestion
    return abs(pilot_velocity_z_max_dn);
```

File: ArduPlane/quadplane.h [hunk 7]
```diff
@@ -154,6 +154,9 @@ class QuadPlane
 
     // return true if the user has set ENABLE
     bool enabled(void) const { return enable != 0; }
+
+    // get value for pilot_velocity_z_max_dn
```
Comment (peterbarker) [inline ArduPlane/quadplane.h:158]: This comment really doesn't add much except clutter :-)

File: ArduPlane/quadplane.cpp [hunk 8]
```diff
@@ -1542,7 +1551,8 @@ float QuadPlane::get_pilot_desired_climb_rate_cms(void) const
     }
     uint16_t dead_zone = plane.channel_throttle->get_dead_zone();
     uint16_t trim = (plane.channel_throttle->get_radio_max() + plane.channel_throttle->get_radio_min())/2;
-    return pilot_velocity_z_max * plane.channel_throttle->pwm_to_angle_dz_trim(dead_zone, trim) / 100.0f;
+    float throttle_request = plane.channel_throttle->pwm_to_angle_dz_trim(dead_zone, trim) / 100.0f;
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:1554]:
```suggestion
    const float throttle_request = plane.channel_throttle->pwm_to_angle_dz_trim(dead_zone, trim) *0.01f;
```

**Review by peterbarker — COMMENTED** [review]
Looks fine to me - but the name question needs to be settled.

File: ArduPlane/quadplane.cpp (hunk 6 above)
Comment (Hwurzburg) [reply ArduPlane/quadplane.cpp]: this suggested change breaks the code....had to add back the brace

File: ArduPlane/quadplane.h (hunk 7 above)
Comment (Hwurzburg) [reply ArduPlane/quadplane.h:158]: agree, but following the pattern...all get functions seem to have a comment header....

Comment (peterbarker) [reply ArduPlane/quadplane.h:158]: Make it useful or remove it IMO.

Comment (Hwurzburg) [reply ArduPlane/quadplane.h]: removed

**Comment by rmackay9** [comment]
I think to be consistent with Copter it would be good to call the parameters PILOT_SPEED_UP and PILOT_SPEED_DN

**Review by tridge — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks @Hwurzburg

**Pull request merged**